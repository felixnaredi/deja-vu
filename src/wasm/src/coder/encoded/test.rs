use std::iter;

use super::*;
use crate::{
  coder::{
    UnseenSetID,
    Version00Coding,
  },
  game_over::GameOver,
};

#[test]
fn encode_decode_same_as_id()
{
  let unseen: Vec<[u8; 1]> = (0..64).map(|x| [x]).collect();

  let game_over = GameOver::new(
    2313308731114687875,
    UnseenSetID::Unspecified,
    unseen.clone(),
    0.4.try_into().unwrap(),
    [Some(1), Some(24), Some(30)],
  );
  let encoded = Version00Coding::encode(&game_over);
  let decoded = Version00Coding::decode(encoded.try_into().unwrap(), unseen).unwrap();

  let mut i = 31;
  for (u0, u1) in iter::zip(game_over.into_iter(), decoded.into_iter()) {
    assert_eq!(u0, u1);
    i -= 1;
  }
  assert_eq!(i, 0);
}

#[test]
fn modifying_checksum_throws_invalid_checksum_error()
{
  let unseen: Vec<[u8; 1]> = (0..64).map(|x| [x]).collect();

  let game_over = GameOver::new(
    2313308731114687875,
    UnseenSetID::Unspecified,
    unseen.clone(),
    0.4.try_into().unwrap(),
    [Some(16), Some(30), Some(34)],
  );
  let mut encoded = Version00Coding::encode(&game_over);

  // xor the checksum with a random value.
  encoded.checksum ^= KNOMUL::permute_index(encoded.checksum, 12228011056065030022);

  assert_eq!(
    EncodedGameOver::try_from(encoded).unwrap_err().to_string(),
    SealedEncodedError::InvalidChecksum.to_string()
  );
}

#[test]
fn modifying_data_throws_invalid_checksum_error()
{
  let unseen: Vec<[u8; 1]> = (0..64).map(|x| [x]).collect();

  let game_over = GameOver::new(
    2313308731114687875,
    UnseenSetID::Unspecified,
    unseen.clone(),
    0.4.try_into().unwrap(),
    [Some(21), Some(46), Some(47)],
  );
  let mut encoded = Version00Coding::encode(&game_over);

  // Swap two letters in the base64 encoded data.

  let (i, j) = (
    KNOMUL::permute_index(encoded.checksum, 15986057670448281097) as usize % encoded.data.len(),
    KNOMUL::permute_index(encoded.checksum, 15772190296526768807) as usize % encoded.data.len(),
  );
  let mut b: Vec<u8> = encoded.data.as_bytes().iter().cloned().collect();
  b.swap(i, j);
  encoded.data = String::from_utf8(b).unwrap();

  assert_eq!(
    EncodedGameOver::try_from(encoded).unwrap_err().to_string(),
    SealedEncodedError::InvalidChecksum.to_string()
  );
}

#[test]
fn invalid_version_throws_unrecognised_version_error()
{
  let unseen: Vec<[u8; 1]> = (0..64).map(|x| [x]).collect();

  let game_over = GameOver::new(
    2313308731114687875,
    UnseenSetID::Unspecified,
    unseen.clone(),
    0.4.try_into().unwrap(),
    [Some(22), Some(42), Some(49)],
  );
  let mut encoded = Version00Coding::encode(&game_over);
  encoded.version = "bad-version".into();

  // The checksum is modified too. This to ensure that version error is prioritized over invalid
  // checksum.
  encoded.checksum = KNOMUL::permute_index(encoded.checksum, 10041457562034272223);

  // The data is modified too. This to ensure that version error is prioritized over invalid
  // data.
  let (i, j) = (
    KNOMUL::permute_index(encoded.checksum, 6813646448085885511) as usize % encoded.data.len(),
    KNOMUL::permute_index(encoded.checksum, 16217660822986183576) as usize % encoded.data.len(),
  );
  let mut b: Vec<u8> = encoded.data.as_bytes().iter().cloned().collect();
  b.swap(i, j);
  encoded.data = String::from_utf8(b).unwrap();

  assert_eq!(
    EncodedGameOver::try_from(encoded).unwrap_err().to_string(),
    SealedEncodedError::UnrecognisedVersion("bad-version".into()).to_string()
  );
}

mod game_over_coder_v01
{
  use std::iter;

  use crate::{
    coder::{
      GameOverCoderV01,
      SealedEncodedGameOver,
      UnseenSetID,
    },
    game_over::GameOver,
  };

  #[test]
  fn encode_decode_equals_id()
  {
    let unseen: Vec<[u8; 1]> = (0..64).map(|x| [x]).collect();
    let game_over = GameOver::new(
      9940370477626720397,
      UnseenSetID::Unspecified,
      unseen.clone(),
      0.4.try_into().unwrap(),
      [Some(9), Some(15), Some(35)],
    );
    let encoded = SealedEncodedGameOver::new::<GameOverCoderV01, _>(&game_over).unwrap();
    let decoded = GameOver::try_from((encoded, unseen)).unwrap();

    assert_eq!(decoded.element_checksum(), game_over.element_checksum());
    assert_eq!(decoded.score(), game_over.score());
    assert_eq!(decoded.lives(), game_over.lives());
    assert!(iter::zip(decoded.into_iter(), game_over.into_iter()).all(|(x, y)| x == y));
  }

  #[test]
  fn detect_use_of_wrong_set_when_decoding()
  {
    let e = SealedEncodedGameOver::new::<GameOverCoderV01, _>(&GameOver::new(
      622451429113938556,
      UnseenSetID::Unspecified,
      // The unseen set used when encoding...
      (0..64).map(|x| [x]).collect(),
      0.4.try_into().unwrap(),
      [Some(30), Some(31), Some(54)],
    ))
    .unwrap();

    assert!(GameOver::try_from((e, (64..128).map(|x| [x]).collect())).is_err());
  }
}
