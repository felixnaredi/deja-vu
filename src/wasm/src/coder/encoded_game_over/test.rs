use super::*;
use crate::{
  coder::{
    UnseenSetID,
    Version00Coding,
  },
  game::SeenThreshold,
  game_over::GameOver,
  rng::{
    IndexedPermutation,
    Konadare192PxPlusPlus,
    KSINK,
  },
};

const S: u64 = 5943847430032560006;

fn generate_game_over(
  s: u64,
  unseen_set_id: Option<UnseenSetID>,
  seen_threshold: Option<SeenThreshold>,
) -> (GameOver<String>, Vec<String>)
{
  let mut rng = Konadare192PxPlusPlus::from_seed(s);

  let mut incorrect = [
    Some(rng.next_with_upper_bound(256) as usize),
    Some(rng.next_with_upper_bound(256) as usize),
    Some(rng.next_with_upper_bound(256) as usize),
  ];
  incorrect.sort();

  let seed = rng.next();

  let unseen: Vec<String> = (0..(256 + rng.next_with_upper_bound(64)))
    .map(|_| {
      base64::encode(
        (0..(8 + rng.next_with_upper_bound(32)))
          .map(|_| rng.next_with_upper_bound(256).try_into().unwrap())
          .collect::<Vec<u8>>()
          .as_slice(),
      )
    })
    .collect();

  (
    GameOver::new(
      seed,
      unseen_set_id.unwrap_or(UnseenSetID::Unspecified),
      unseen.clone(),
      seen_threshold.unwrap_or_else(|| {
        (rng.next() as f64 / ((1u64 << 63) | ((1u64 << 63) - 1)) as f64)
          .try_into()
          .unwrap()
      }),
      incorrect,
    ),
    unseen,
  )
}

#[test]
fn f()
{
  println!(
    "{}",
    KSINK::hash(S, "2202 TSEC 85:62:71 32 tcO nuS".as_bytes())
  )
}

#[cfg(test)]
mod version00coding
{
  use super::*;

  #[test]
  fn encode_decode_same_as_id()
  {
    let (game_over, unseen) = generate_game_over(
      542338303675782954,
      Some(UnseenSetID::DictionaryFr01),
      Some(0.4.try_into().unwrap()),
    );
    let encoded = SealedEncodedGameOver::new::<Version00Coding, _>(&game_over).unwrap();
    let decoded: GameOver<String> = (encoded, unseen).try_into().unwrap();
    assert_eq!(decoded, game_over);
  }

  #[test]
  fn encode_outputs_expected_literal()
  {
    let (game_over, _) = generate_game_over(
      10607240176095180561,
      Some(UnseenSetID::DictionaryFr01),
      Some(0.4.try_into().unwrap()),
    );

    assert_eq!(serde_urlencoded::to_string(
      SealedEncodedGameOver::new::<Version00Coding, _>(&game_over).unwrap()
    )
    .unwrap(),
    "version=00&checksum=15367392902022108743&data=eyJ1bnNlZW5faWQiOiJEaWN0aW9uYXJ5RnIwMSIsInNlZWQiOjc3NzAzNzYyMDA2OTQ2Mjc5MTcsImluY29ycmVjdF9jb21taXRzIjpbMzMsNDksMTIxXX0%3D&unseen_set_id=DictionaryFr01")
  }

  #[test]
  fn decode_literal_outputs_expected_game_over()
  {
    let s = "version=00&checksum=15965453229032169769&data=eyJ1bnNlZW5faWQiOiJEaWN0aW9uYXJ5RnIwMSIsInNlZWQiOjk3NTIwMTE1MzkzNzAxMjE1MDgsImluY29ycmVjdF9jb21taXRzIjpbMTEsMTI2LDI0OV19&unseen_set_id=DictionaryFr01";

    let (game_over, unseen) = generate_game_over(
      12837024807937892714,
      Some(UnseenSetID::DictionaryFr01),
      Some(0.4.try_into().unwrap()),
    );

    let decoded: GameOver<String> = (serde_urlencoded::from_str(s).unwrap(), unseen)
      .try_into()
      .unwrap();
    assert_eq!(decoded, game_over);
  }

  #[test]
  fn decode_literal_with_corrupt_checksum()
  {
    let (_, unseen) = generate_game_over(
      9354512038846019256,
      Some(UnseenSetID::DictionaryFr01),
      Some(0.4.try_into().unwrap()),
    );

    let s = "version=00&checksum=114325682255183593&data=eyJ1bnNlZW5faWQiOiJEaWN0aW9uYXJ5RnIwMSIsInNlZWQiOjE2MTM3Mjg1Nzg3MTIwODk2OTc0LCJpbmNvcnJlY3RfY29tbWl0cyI6Wzc2LDc3LDE2MF19&unseen_set_id=DictionaryFr01";
    assert!(matches!(
      TryInto::<GameOver<String>>::try_into((
        serde_urlencoded::from_str::<SealedEncodedGameOver>(s).unwrap(),
        unseen
      )),
      Err(_)
    ));
  }

  #[test]
  fn decode_literal_with_corrupt_data()
  {
    let (_, unseen) = generate_game_over(
      3088894222515458658,
      Some(UnseenSetID::DictionaryFr01),
      Some(0.4.try_into().unwrap()),
    );

    let s = "version=00&checksum=15649906318563636177&data=eyJ1bnNlZW5faWQiOiJEaWN0aW9uYXJ5RnIwMSIsInNlZWQiOjE5MzUyNTQ2NDY0MTkzOTI1NzYsImluY29ycmVjdF9jb21taXRzIjpbMTQsMTUsMTUzXX0%3D&unseen_set_id=DictionaryFr01";
    assert!(matches!(
      TryInto::<GameOver<String>>::try_into((
        serde_urlencoded::from_str::<SealedEncodedGameOver>(s).unwrap(),
        unseen
      )),
      Err(_)
    ));
  }

  #[test]
  fn decode_url_with_missing_fields()
  {
    let (game_over, unseen) = generate_game_over(
      17561082018945952039,
      Some(UnseenSetID::DictionaryFr01),
      Some(0.4.try_into().unwrap()),
    );

    let version = "version=00";
    let checksum = "checksum=16916324704992223141";
    let data = "data=eyJ1bnNlZW5faWQiOiJEaWN0aW9uYXJ5RnIwMSIsInNlZWQiOjcxMzA5MDM0Nzg0NDc2NzAwMjUsImluY29ycmVjdF9jb21taXRzIjpbMTMxLDE3NiwyNDFdfQ%3D%3D";
    let unseen_set_id = "unseen_set_id=DictionaryFr01";

    for i in 0..15 {
      let fields: Vec<String> = [
        (i & 1 != 0).then(|| version),
        (i & 2 != 0).then(|| checksum),
        (i & 4 != 0).then(|| data),
        (i & 8 != 0).then(|| unseen_set_id),
      ]
      .iter()
      .filter_map(|x| x.map(|x| x.into()))
      .collect();
      let fields = fields.join("&");

      if i == 7 {
        // All fields but `unseen_set_id` is present. This is a valid URL.
        let decoded: GameOver<String> =
          (serde_urlencoded::from_str(&fields).unwrap(), unseen.clone())
            .try_into()
            .unwrap();
        assert_eq!(decoded, game_over);
      } else {
        // Invalid URL. This will be detected when deserializing the URL into a
        // `SealedEncodedGameOver`
        assert!(matches!(
          serde_urlencoded::from_str::<SealedEncodedGameOver>(&fields),
          Err(_)
        ));
      }
    }
  }
}

/*
mod game_over_coder_v01
{
  use std::iter;

  use crate::{
    coder::{
      GameOverCoderV01,
      SealedEncodedGameOver,
      UnseenSetID,
    },
    game_over::GameOver,
  };

  #[test]
  fn encode_decode_equals_id()
  {
    let unseen: Vec<[u8; 1]> = (0..64).map(|x| [x]).collect();
    let game_over = GameOver::new(
      9940370477626720397,
      UnseenSetID::Unspecified,
      unseen.clone(),
      0.4.try_into().unwrap(),
      [Some(9), Some(15), Some(35)],
    );
    let encoded = SealedEncodedGameOver::new::<GameOverCoderV01, _>(game_over.clone()).unwrap();
    let decoded = GameOver::try_from((encoded, unseen)).unwrap();

    assert_eq!(decoded.element_checksum(), game_over.element_checksum());
    assert_eq!(decoded.score(), game_over.score());
    assert_eq!(decoded.lives(), game_over.lives());
    assert!(iter::zip(decoded.into_iter(), game_over.into_iter()).all(|(x, y)| x == y));
  }

  #[test]
  fn detect_use_of_wrong_set_when_decoding()
  {
    let e = SealedEncodedGameOver::new::<GameOverCoderV01, _>(GameOver::new(
      622451429113938556,
      UnseenSetID::Unspecified,
      // The unseen set used when encoding...
      (0..64).map(|x| [x]).collect(),
      0.4.try_into().unwrap(),
      [Some(30), Some(31), Some(54)],
    ))
    .unwrap();

    assert!(GameOver::try_from((e, (64..128).map(|x| [x]).collect())).is_err());
  }
}
*/
