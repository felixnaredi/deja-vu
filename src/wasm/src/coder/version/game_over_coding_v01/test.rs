#[cfg(test)]
mod temp
{
  use crate::rng::{
    IndexedPermutation,
    Konadare192PxPlusPlus,
    KSINK,
  };
  #[test]
  fn f()
  {
    // 2123399681707578173
    // 17792727522076118826
    // 3419954961163653668
    println!(
      "{}",
      KSINK::permute_index(3419954961163653668, 17792727522076118826)
    );

    panic!("REMOVE THIS TEST BEFORE PR")
  }

  #[test]
  fn g()
  {
    let mut rng = Konadare192PxPlusPlus::from_seed(8366408846780968924);
    let mut s: Vec<Option<usize>> = (0..3)
      .map(|_| rng.next_with_upper_bound(64) as usize)
      .map(|x| Some(x))
      .collect();
    s.sort();
    println!("{:?}", s);

    // [Some(30), Some(31), Some(54)]
    panic!("REMOVE THIS TEST BEFORE PR")
  }
}

use std::iter;

use super::GameOverCodingV01;
use crate::{
  coder::{
    unseen_set_id::UnseenSetID,
    version::game_over_coding_v01::GameOverCodingV01Error,
  },
  game_over::GameOver,
};

#[test]
fn encode_decode_equals_id()
{
  let unseen: Vec<[u8; 1]> = (0..64).map(|x| [x]).collect();
  let game_over = GameOver::new(
    9940370477626720397,
    UnseenSetID::Unspecified,
    unseen.clone(),
    0.4.try_into().unwrap(),
    [Some(9), Some(15), Some(35)],
  );
  let encoded = GameOverCodingV01::encode(game_over.clone(), UnseenSetID::Unspecified).unwrap();
  let decoded = GameOverCodingV01::decode(encoded.try_into().unwrap(), unseen).unwrap();

  assert_eq!(decoded.element_checksum(), game_over.element_checksum());
  assert_eq!(decoded.score(), game_over.score());
  assert_eq!(decoded.lives(), game_over.lives());
  assert!(iter::zip(decoded.into_iter(), game_over.into_iter()).all(|(x, y)| x == y));
}

#[test]
fn detect_use_of_wrong_set_when_encoding()
{
  let game_over = GameOver::new(
    13206503794884972104,
    UnseenSetID::Unspecified,
    (0..64).map(|x| [x]).collect(),
    0.4.try_into().unwrap(),
    [Some(4), Some(33), Some(45)],
  );
  assert!(matches!(
    GameOverCodingV01::encode(game_over.clone(), UnseenSetID::Unspecified,),
    Err(GameOverCodingV01Error::BadElementChecksum)
  ));
}

#[test]
fn detect_use_of_wrong_set_when_decoding()
{
  let game_over = GameOver::new(
    622451429113938556,
    UnseenSetID::Unspecified,
    (0..64).map(|x| [x]).collect(),
    0.4.try_into().unwrap(),
    [Some(30), Some(31), Some(54)],
  );
  let encoded = GameOverCodingV01::encode(game_over.clone(), UnseenSetID::Unspecified).unwrap();
  assert!(matches!(
    GameOverCodingV01::decode(
      encoded.try_into().unwrap(),
      (64..128).map(|x| [x]).collect(),
    ),
    Err(GameOverCodingV01Error::BadElementChecksum)
  ));
}

#[test]
fn decode_detect_bad_element_checksum()
{
  panic!("not implemented");
}
