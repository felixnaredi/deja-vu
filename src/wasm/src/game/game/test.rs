use std::collections::{
  HashSet,
  LinkedList,
};

use super::*;
use crate::rng::{
  IndexedPermutation,
  KSINK,
};

#[test]
fn commit_errors_when_empty()
{
  let mut game = Game::new(
    10539,
    0.4.try_into().unwrap(),
    (0..1).map(|x| [x]).collect(),
  );
  assert_eq!(game.commit_seen(), Err(GameError::EmptyCommit));
  assert_eq!(game.commit_unseen(), Err(GameError::EmptyCommit));
}

#[test]
fn next_throws_errors_with_uncommited_results()
{
  let mut game = Game::new(
    11484,
    0.0.try_into().unwrap(),
    (0..2).map(|x| [x]).collect(),
  );
  assert!(matches!(game.next(), Ok(_)));
  assert_eq!(game.next(), Err(GameError::NextCalledWithUncommitedResult));

  let mut game = Game::new(
    11898,
    1.0.try_into().unwrap(),
    (0..2).map(|x| [x]).collect(),
  );
  assert!(matches!(game.next(), Ok(_)));
  assert_eq!(game.next(), Err(GameError::NextCalledWithUncommitedResult));
}

#[test]
fn three_strikes_guessing_seen_causes_game_over()
{
  let mut game = Game::new(
    12584,
    0.0.try_into().unwrap(),
    (0..4).map(|x| [x]).collect(),
  );
  for _ in 0..3 {
    assert!(matches!(game.next(), Ok(_)));
    assert!(matches!(game.commit_seen(), Ok(false)));
  }
  assert_eq!(game.next(), Err(GameError::GameOver));
  assert_eq!(game.commit_seen(), Err(GameError::GameOver));
  assert_eq!(game.commit_unseen(), Err(GameError::GameOver));
}

#[test]
fn three_strikes_guessing_unseen_causes_game_over()
{
  let mut game = Game::new(
    12554,
    1.0.try_into().unwrap(),
    (0..4).map(|x| [x]).collect(),
  );

  // First two `next` will always be unseen.
  assert!(matches!(game.next(), Ok(_)));
  assert!(matches!(game.commit_unseen(), Ok(true)));
  assert!(matches!(game.next(), Ok(_)));
  assert!(matches!(game.commit_unseen(), Ok(true)));

  for _ in 0..3 {
    assert!(matches!(game.next(), Ok(_)));
    assert!(matches!(game.commit_unseen(), Ok(false)));
  }
  assert_eq!(game.next(), Err(GameError::GameOver));
  assert_eq!(game.commit_seen(), Err(GameError::GameOver));
  assert_eq!(game.commit_unseen(), Err(GameError::GameOver));
}

#[test]
fn next_unseen_returns_unique_and_errors_when_empty()
{
  let n = 16;
  let mut game = Game::new(6237, 0.0.try_into().unwrap(), (0..n).map(|x| [x]).collect());
  let mut s = HashSet::new();

  for _ in 0..n {
    if !s.insert(game.next().unwrap().clone()) {
      panic!("`next_unseen` generated already generated value")
    }
    assert!(game.commit_unseen().unwrap());
  }

  assert_eq!(game.next(), Err(GameError::UnseenEmpty));
}

#[test]
fn next_seen_returns_error_when_too_few_elements()
{
  let mut game = Game::new(8833, 1.0.try_into().unwrap(), (0..0).map(|x| [x]).collect());
  assert!(matches!(game.next(), Err(_)));

  let mut game = Game::new(
    19119,
    1.0.try_into().unwrap(),
    (0..1).map(|x| [x]).collect(),
  );
  assert!(matches!(game.next(), Ok(_)));
  assert!(matches!(game.next(), Err(_)));
}

#[test]
fn next_generates_equal_output_for_equal_input()
{
  let mut game1 = Game::new(
    10335,
    0.5.try_into().unwrap(),
    (0..16).map(|x| [x]).collect(),
  );
  let mut game2 = Game::new(
    10335,
    0.5.try_into().unwrap(),
    (0..16).map(|x| [x]).collect(),
  );

  for y in [true, true, true, false, false, false, true, true, true] {
    assert_eq!(game1.next(), game2.next());
    if y {
      assert_eq!(game1.commit_seen(), game2.commit_seen());
    } else {
      assert_eq!(game1.commit_unseen(), game2.commit_unseen());
    }
  }
}

#[test]
fn next_never_generates_same_twice_in_a_row()
{
  let mut game = Game::new(
    10335,
    1.0.try_into().unwrap(),
    (0..4).map(|x| [x]).collect(),
  );
  let mut previous = game.next().unwrap().clone();
  for _ in 0..16 {
    game.commit_seen().unwrap();
    let x = game.next().unwrap().clone();
    assert_ne!(previous, x);
    previous = x;
  }
}

#[test]
fn score_increases_on_correct_commit()
{
  let mut game = Game::new(
    11976,
    0.5.try_into().unwrap(),
    (0..8).map(|x| [x]).collect(),
  );
  for (score, guess_seen) in [
    (1, false),
    (2, false),
    (2, true),
    (3, false),
    (4, true),
    (4, false),
  ] {
    game.next().unwrap();

    if guess_seen {
      game.commit_seen().unwrap();
    } else {
      game.commit_unseen().unwrap();
    }

    assert_eq!(game.score(), score);
  }
}

#[test]
fn life_decrease_on_incorrect_commit()
{
  let mut game = Game::new(
    211391,
    0.5.try_into().unwrap(),
    (0..8).map(|x| [x]).collect(),
  );
  for (lives, guess_seen) in [
    (3, false),
    (3, false),
    (3, true),
    (3, false),
    (2, true),
    (2, true),
    (2, false),
    (1, true),
    (1, false),
    (0, false),
  ] {
    game.next().unwrap();

    if guess_seen {
      game.commit_seen().unwrap();
    } else {
      game.commit_unseen().unwrap();
    }
    assert_eq!(game.lives(), lives);
  }

  assert!(game.finished());
}

#[test]
fn indices_are_same_as_incorrect_commits()
{
  let mut game = Game::new(
    877326994,
    0.5.try_into().unwrap(),
    (0..16).map(|x| [x]).collect(),
  );
  let guess_seen = [2, 3, 5, 6, 11];
  let wrongs = [3, 8, 11];

  for i in 0..12 {
    game.next().unwrap();

    if guess_seen.contains(&i) {
      assert!(game.commit_seen().unwrap() == !wrongs.contains(&i));
    } else {
      assert!(game.commit_unseen().unwrap() == !wrongs.contains(&i));
    }
  }

  assert_eq!(
    wrongs
      .iter()
      .map(|x| Some(x.clone()))
      .collect::<Vec<Option<usize>>>(),
    game.incorrect_commits()
  );
}

#[test]
fn after_reseting_a_game_it_produces_same_output_given_same_input()
{
  let mut game = Game::new(
    6314949274223572360,
    0.4.try_into().unwrap(),
    (0..32).map(|x| [x]).collect(),
  );
  let guess_seen = [4, 5, 8, 12, 17, 18, 19, 20, 21, 22, 24, 25];

  let mut elements = LinkedList::new();
  let mut i = 0;
  while !game.finished() {
    elements.push_back(game.next().unwrap().clone());
    if guess_seen.contains(&i) {
      game.commit_seen().unwrap();
    } else {
      game.commit_unseen().unwrap();
    }
    i += 1;
  }

  let incorrect = game.incorrect_commits().clone();

  game.reset();
  let mut i = 0;
  while !game.finished() {
    assert_eq!(game.next().unwrap(), &elements.pop_front().unwrap());
    if guess_seen.contains(&i) {
      game.commit_seen().unwrap();
    } else {
      game.commit_unseen().unwrap();
    }
    i += 1;
  }

  assert!(elements.is_empty());
  assert_eq!(game.incorrect_commits(), incorrect);
}

#[test]
fn element_checksum_outputs_as_expected()
{
  let mut game = Game::new(
    9004491465204253423,
    0.3.try_into().unwrap(),
    vec![
      "4hYKvDtHGSo=",
      "/QaJrz+CL/Q=",
      "d6z9AveYhgY=",
      "3t69Oqkp/DM=",
      "1/sKN1Lxhb8=",
      "MXe8R8hcGds=",
      "ZOIvjnNgSgo=",
      "XF1nMT3+sNI=",
      "l5a/lq691wc=",
      "5ogqZBBRKAc=",
      "PMVOmOlhnRc=",
      "7jAprh4MM34=",
      "M1q/CZxTsMU=",
      "vGMrB1HuhnM=",
      "SPYz7ePBqJU=",
      "8O/quTDz7zw=",
    ],
  );

  let mut s = DEFAULT_ELEMENT_CHECKSUM;
  assert_eq!(s, game.element_checksum());

  let mut m = 0;

  while !game.finished() {
    s = KSINK::hash(s, game.next().unwrap().as_bytes());
    assert_eq!(s, game.element_checksum());

    m += 1;
    m %= 3;
    if m == 0 {
      game.commit_seen().unwrap();
    } else {
      game.commit_unseen().unwrap();
    }
  }
}
