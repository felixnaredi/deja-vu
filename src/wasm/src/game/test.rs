#[cfg(test)]
use std::collections::HashSet;

#[allow(unused_imports)]
use super::*;

#[test]
fn commit_errors_when_empty()
{
  let mut game = Game::new(10539, 0.4, 0..1);
  assert_eq!(game.commit_seen(), Err(GameError::EmptyCommit));
  assert_eq!(game.commit_unseen(), Err(GameError::EmptyCommit));
}

#[test]
fn next_throws_errors_with_uncommited_results()
{
  let mut game = Game::new(11484, 0.0, 0..2);
  assert!(matches!(game.next(), Ok(_)));
  assert_eq!(game.next(), Err(GameError::NextCalledWithUncommitedResult));

  let mut game = Game::new(11898, 1.0, 0..2);
  assert!(matches!(game.next(), Ok(_)));
  assert_eq!(game.next(), Err(GameError::NextCalledWithUncommitedResult));
}

#[test]
fn three_strikes_guessing_seen_causes_game_over()
{
  let mut game = Game::new(12584, 0.0, 0..4);
  for _ in 0..3 {
    assert!(matches!(game.next(), Ok(_)));
    assert!(matches!(game.commit_seen(), Ok(false)));
  }
  assert_eq!(game.next(), Err(GameError::GameOver));
  assert_eq!(game.commit_seen(), Err(GameError::GameOver));
  assert_eq!(game.commit_unseen(), Err(GameError::GameOver));
}

#[test]
fn three_strikes_guessing_unseen_causes_game_over()
{
  let mut game = Game::new(12554, 1.0, 0..4);

  // First two `next` will always be unseen.
  assert!(matches!(game.next(), Ok(_)));
  assert!(matches!(game.commit_unseen(), Ok(true)));
  assert!(matches!(game.next(), Ok(_)));
  assert!(matches!(game.commit_unseen(), Ok(true)));

  for _ in 0..3 {
    assert!(matches!(game.next(), Ok(_)));
    assert!(matches!(game.commit_unseen(), Ok(false)));
  }
  assert_eq!(game.next(), Err(GameError::GameOver));
  assert_eq!(game.commit_seen(), Err(GameError::GameOver));
  assert_eq!(game.commit_unseen(), Err(GameError::GameOver));
}

#[test]
fn next_unseen_returns_unique_and_errors_when_empty()
{
  let n = 16;
  let mut game = Game::new(6237, 0.0, 0..n);
  let mut s = HashSet::new();

  for _ in 0..n {
    if !s.insert(game.next().unwrap().clone()) {
      panic!("`next_unseen` generated already generated value")
    }
    assert!(game.commit_unseen().unwrap());
  }

  assert_eq!(game.next(), Err(GameError::UnseenEmpty));
}

#[test]
fn next_seen_returns_error_when_too_few_elements()
{
  let mut game = Game::new(8833, 1.0, 0..0);
  assert!(matches!(game.next(), Err(_)));

  let mut game = Game::new(19119, 1.0, 0..1);
  assert!(matches!(game.next(), Ok(_)));
  assert!(matches!(game.next(), Err(_)));
}

#[test]
fn next_generates_equal_output_for_equal_input()
{
  let mut game1 = Game::new(10335, 0.5, 0..16);
  let mut game2 = Game::new(10335, 0.5, 0..16);

  for y in [true, true, true, false, false, false, true, true, true] {
    assert_eq!(game1.next(), game2.next());
    if y {
      assert_eq!(game1.commit_seen(), game2.commit_seen());
    } else {
      assert_eq!(game1.commit_unseen(), game2.commit_unseen());
    }
  }
}

#[test]
fn next_never_generates_same_twice_in_a_row()
{
  let mut game = Game::new(10335, 1.0, 0..4);
  let mut previous = game.next().unwrap().clone();
  for _ in 0..16 {
    game.commit_seen().unwrap();
    let x = game.next().unwrap().clone();
    assert_ne!(previous, x);
    previous = x;
  }
}
