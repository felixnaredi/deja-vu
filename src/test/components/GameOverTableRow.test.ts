import { generateCommit } from "@/test/commit-mock";

test("temp0", () => {
  console.log(generateCommit({ seed: BigInt(100) }).element());
});

test("temp1", () => {
  console.log(
    generateCommit({ seed: BigInt(100) })
      .actual()
      .isSeen()
  );
});
