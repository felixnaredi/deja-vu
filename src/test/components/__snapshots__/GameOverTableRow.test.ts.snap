// Vitest Snapshot v1

exports[`snapshots > guessing 'seen' when the actual is 'seen' 1`] = `
"<div class=\\"grid grid-cols-8 gap-4 border-b-2\\">
  <div class=\\"h-full justify-self-center\\">
    <eye-symbol-stub class=\\"h-full w-5\\"></eye-symbol-stub>
  </div>
  <div class=\\"col-span-5 ml-4\\">7yEb1GA6Xaj8J7e5D</div>
  <div class=\\"justify-self-center\\">
    <div class=\\"h-full\\">
      <tick-symbol-stub class=\\"w-5 h-full\\"></tick-symbol-stub>
    </div>
  </div>
  <div></div>
</div>"
`;

exports[`snapshots > guessing 'seen' when the actual is 'unseen' 1`] = `
"<div class=\\"grid grid-cols-8 gap-4 border-b-2 bg-red-50\\">
  <div></div>
  <div class=\\"col-span-5 ml-4\\">sPvi1nq9VRMiQ</div>
  <div class=\\"justify-self-center\\">
    <div class=\\"h-full\\">
      <cross-symbol-stub class=\\"w-5 h-full\\"></cross-symbol-stub>
    </div>
  </div>
  <div></div>
</div>"
`;

exports[`snapshots > guessing 'unseen' when the actual is 'seen' 1`] = `
"<div class=\\"grid grid-cols-8 gap-4 border-b-2 bg-red-50\\">
  <div class=\\"h-full justify-self-center\\">
    <eye-symbol-stub class=\\"h-full w-5\\"></eye-symbol-stub>
  </div>
  <div class=\\"col-span-5 ml-4\\">Txlxug3V8T9ClBZd</div>
  <div class=\\"justify-self-center\\">
    <div></div>
  </div>
  <div class=\\"h-full justify-self-center\\">
    <cross-symbol-stub class=\\"w-5 h-full\\"></cross-symbol-stub>
  </div>
</div>"
`;

exports[`snapshots > guessing 'unseen' when the actual is 'unseen' 1`] = `
"<div class=\\"grid grid-cols-8 gap-4 border-b-2\\">
  <div></div>
  <div class=\\"col-span-5 ml-4\\">Z54WuMeMC7fXYq5NTVht</div>
  <div class=\\"justify-self-center\\">
    <div></div>
  </div>
  <div class=\\"h-full justify-self-center\\">
    <tick-symbol-stub class=\\"w-5 h-full\\"></tick-symbol-stub>
  </div>
</div>"
`;
